Enable multiple function composition in prefix form

JeffFessler opened the pull request:
Document the prefix form and extend the prefix form to support composition of 3 or more functions.
If there is favorable sentiment for this then I can also add tests.
There was discussion in 2016 about related issues but I don't see anything more recent
https://github.com/JuliaLang/julia/pull/17184

JeffFessler committed:
Enable multiple function composition in prefix form

Document the prefix form and extend the prefix form to support composition of 3 or more functions.
If there is favorable sentiment for this then I can also add tests.
There was discussion in 2016 about related issues but I don't see anything more recent
https://github.com/JuliaLang/julia/pull/17184

JeffFessler committed:
Refine multiple function composition

Thanks for the feedback.  Done.
Also added a single doctest that should suffice for codecov.

JeffFessler commented on base/operators.jl:
```diff
@@ -838,6 +842,9 @@ julia> map(uppercase∘first, ["apple", "banana", "carrot"])
 ∘(f, g) = (x...)->f(g(x...))
 
+∘(f::Function, g::Function, h::Function) = ∘(f, ∘(g, h))
+∘(f::Function, g::Function, h::Function, is::Function...) = ∘(∘(f, g), h, is...)
```
Good points, thanks. I made those changes and added a doctest.

dkarrasch commented on base/operators.jl:
```diff
@@ -827,17 +827,28 @@ julia> [1:5;] |> x->x.^2 |> sum |> inv
 
+Function composition also works in prefix form: `∘(f,g)` is the same as `f ∘ g`.
+The prefix form also supports multiple function composition: `∘(f,g,h) = f ∘ g ∘ h`
+and splatting for an iterable with functions `∘(fs...)`
```
The sentence should finish with a period, I guess.

JeffFessler replied:
Fixed.

dkarrasch commented on base/operators.jl:
```diff
@@ -827,17 +827,28 @@ julia> [1:5;] |> x->x.^2 |> sum |> inv
+∘(f, g, h) = ∘(f ∘ g, h)
+∘(f, g, h, is...) = ∘(∘(f, g), h, is...)
+
+
```
Maybe one of the empty lines can be removed: there used to be two empty lines, now there are 3. I'm not sure, but the two new definitions belong to the two-function case, so maybe there's no need for the empty line there?

JeffFessler replied:
Yep.  So happy that the main issues are spacing :)

JeffFessler committed:
Update punctuation and spacing.

I also removed "for functions" because composition can work more generally it seems.

dkarrasch commented on base/operators.jl:
```diff
@@ -827,17 +827,25 @@ julia> [1:5;] |> x->x.^2 |> sum |> inv
 entered in the Julia REPL (and most editors, appropriately configured) by typing `\\circ<tab>`.
 
+Function composition also works in prefix form: `∘(f,g)` is the same as `f ∘ g`.
+The prefix form also supports multiple function composition: `∘(f,g,h) = f ∘ g ∘ h`
```
I'm not a native speaker, but how does "supports composition of multiple functions" sound?

StefanKarpinski replied:
```suggestion
The prefix form also supports composition of multiple functions: `∘(f,g,h) = f ∘ g ∘ h`
```

dkarrasch reviewed (commented):
Since this is a new feature, I guess we will need to announce it in a one-liner in NEWS?

StefanKarpinski commented on base/operators.jl:
```diff
@@ -827,17 +827,25 @@ julia> [1:5;] |> x->x.^2 |> sum |> inv
 ∘(f, g) = (x...)->f(g(x...))
-
+∘(f, g, h) = ∘(f ∘ g, h)
+∘(f, g, h, is...) = ∘(∘(f, g), h, is...)
```
Can't this be done with just a single method?
```jl
∘(f, g, h...) = ∘(f ∘ g, h...)
```
Edit: confirmed, it can.

JeffFessler replied:
I had been hoping for a one-liner and I had tried that in a test sandbox but I kept getting infinite recursion, but I see now that I had errors in the way I was using `Base.:(∘)`
It is much more elegant as a single line so I have updated the PR to use that version.  Thanks!

StefanKarpinski commented on base/operators.jl:
```diff
@@ -827,17 +827,25 @@ julia> [1:5;] |> x->x.^2 |> sum |> inv
  'B'
  'C'
+
+julia> fs = (x -> 2x, x -> x/2, x -> x-1, x -> x+1); f = ∘(fs...); f(3) == 3
```
I would expand this out so that it's easier to follow the interaction:
```suggestion
julia> fs = [
           x -> 2x
           x -> x/2
           x -> x-1
           x -> x+1
       ];

julia> ∘(fs...)(3)
3.0
```

StefanKarpinski commented on base/operators.jl:
```diff
@@ -827,17 +827,25 @@ julia> [1:5;] |> x->x.^2 |> sum |> inv
 Compose functions: i.e. `(f ∘ g)(args...)` means `f(g(args...))`. The `∘` symbol can be
 entered in the Julia REPL (and most editors, appropriately configured) by typing `\\circ<tab>`.
 
+Function composition also works in prefix form: `∘(f,g)` is the same as `f ∘ g`.
```
```suggestion
Function composition also works in prefix form: `∘(f, g)` is the same as `f ∘ g`.
```

StefanKarpinski commented on base/operators.jl:
```diff
@@ -827,17 +827,25 @@ julia> [1:5;] |> x->x.^2 |> sum |> inv
 entered in the Julia REPL (and most editors, appropriately configured) by typing `\\circ<tab>`.
 
+Function composition also works in prefix form: `∘(f,g)` is the same as `f ∘ g`.
+The prefix form also supports multiple function composition: `∘(f,g,h) = f ∘ g ∘ h`
```
```suggestion
The prefix form also supports multiple function composition: `∘(f, g, h) = f ∘ g ∘ h`
```

StefanKarpinski replied:
Bump. This one too.

JeffFessler replied:
done

StefanKarpinski commented:
Nice addition! Makes perfect sense that composition should work for arbitrary args.

JeffFessler committed:
Reword comment

Co-Authored-By: Stefan Karpinski <stefan@karpinski.org>

JeffFessler committed:
Expand doctest

I was trying to minimize number of lines changed :)

Co-Authored-By: Stefan Karpinski <stefan@karpinski.org>

JeffFessler committed:
Simplify to a single new line

JeffFessler committed:
Add space

Co-Authored-By: Stefan Karpinski <stefan@karpinski.org>

JeffFessler committed:
More spacing in docstring

StefanKarpinski commented on base/operators.jl:
```diff
@@ -827,17 +827,32 @@ julia> [1:5;] |> x->x.^2 |> sum |> inv
 entered in the Julia REPL (and most editors, appropriately configured) by typing `\\circ<tab>`.
 
+Function composition also works in prefix form: `∘(f, g)` is the same as `f ∘ g`.
+The prefix form also supports composition of multiple functions: `∘(f, g, h) = f ∘ g ∘ h`
```
```suggestion
The prefix form supports composition of multiple functions: `∘(f, g, h) = f ∘ g ∘ h`
```
Getting really nitpicky now—avoid repetition of `also`.

JeffFessler committed:
Cut also

Co-Authored-By: Stefan Karpinski <stefan@karpinski.org>

StefanKarpinski commented on base/operators.jl:
```diff
@@ -827,17 +827,32 @@ julia> [1:5;] |> x->x.^2 |> sum |> inv
 
+Function composition also works in prefix form: `∘(f, g)` is the same as `f ∘ g`.
+The prefix form supports composition of multiple functions: `∘(f, g, h) = f ∘ g ∘ h`
+and splatting `∘(fs...)` for an iterable.
```
```suggestion
and splatting `∘(fs...)` for composing an iterable collection of functions.
```

JeffFessler committed:
Refine docstring

I was avoiding the term "function" because an earlier version had unnecessary `Function` typing but I agree that the description is more clear this way.

Co-Authored-By: Stefan Karpinski <stefan@karpinski.org>

StefanKarpinski commented:
Looking good. Next round of improvements:
1. A couple of simple tests (although there's already the doc test)
2. Add a `!!!compat "Julia 1.4"` annotation in the docstring indicating that this is a new feature
3. Add an entry in NEWS linking to this PR and briefly describing the feature.

JeffFessler committed:
Multiple function composition 

per https://github.com/JuliaLang/julia/pull/33568

JeffFessler committed:
Add compat 1.4 to docstring

JeffFessler committed:
Add tests for multiple function composition

JeffFessler committed:
Refine wording

JeffFessler committed:
Fix doctest

spurious `true` left from older version of doctest

JeffFessler committed:
Cut superfluous ()

JeffFessler commented:
I've done the "next round of improvements" list.  Just made some last tweaks...

dkarrasch commented:
A bit off-topic: @StefanKarpinski I saw you dismissed my earlier review, which was referring to a much earlier state of the PR. Should I have actively changed my review assessment as to not confuse people? Obviously, it was not meant to "request changes" indefinitely. Can I take back a "request changes" without necessarily "approving" the PR?

fredrikekre commented:
This PR failed CI due to trailing whitespace.

StefanKarpinski commented:
Hmm. I'm not sure. I think you can mark individual comments as resolved but I'm not sure how you mark a review as resolved other than what I did — i.e. dismissing it, which felt a bit, uh, dismissive, but seemed like the only thing to do in order to let the PR be mergeable. Maybe leaving a "comment" review puts it into a neither approved nor denied state?

StefanKarpinski commented:
It's really annoying that GitHub will show you the deletion of trailing whitespace in a diff but won't show you the insertion of trailing whitespace, so it's impossible to tell by reviewing a diff that it introduces trailing whitespace. Sorry I missed the whitespace failure in CI. I'm so used to a lone failing CI test for simple changes like being inconsequential that I just ignored it out of habit.

JeffFessler commented:
Thanks everyone - one line of new code and two dozen lines of comments and tests :)

StefanKarpinski commented:
That's the way it goes 😁

jw3126 commented:
I think we should also have
```julia
∘() = identity
∘(f) = f
```

StefanKarpinski commented:
Seems like a good addition. Care to make a PR, @jw3126?

jw3126 commented:
Sure thing!
